### PR TITLE
Reduce contention in scheduler

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

**Comment by erichwang** [comment]
looks good

**Comment by dain** [comment]
Applied all of Eric's comments.  Can you take a look @martint ?

**Comment by dain** [comment]
@martint updated.  Can you review again.

**Comment by dain** [comment]
Updated

File: presto-main/src/main/java/com/facebook/presto/server/ExponentialBackOff.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    public synchronized long getFailureCount()
+    {
+        return failureCount;
+    }
+
+    public synchronized Duration getTimeSinceLastSuccess()
+    {
+        // read lastSuccessfulRequest before getting current time to avoid race conditions
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/server/ExponentialBackOff.java]: this comment no longer applies

File: presto-main/src/main/java/com/facebook/presto/server/ExponentialBackOff.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    public synchronized void success()
+    {
+        lastSuccessTime = ticker.read();
+        failureCount = 0;
+        lastFailureTime = 0;
+    }
+
+    public synchronized boolean failure()
+    {
+        // read lastSuccessfulRequest before getting current time to avoid race conditions
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/server/ExponentialBackOff.java]: same here

File: presto-main/src/main/java/com/facebook/presto/server/HttpRemoteTask.java [hunk 3]
```diff
@@ -821,35 +829,35 @@ public void onFailure(Throwable t)
...
-            // don't update too fast in the face of errors
-            if (errorCount.get() > 0) {
-                errorRequestRateLimiter.acquire();
+            long delayNanos = backoff.getBackoffDelayNanos();
+
+            if (delayNanos == 0) {
+                return Futures.immediateFuture(null);
             }
+
+            SettableFuture<Object> future = SettableFuture.create();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRemoteTask.java]: Use FutureTask (seems my comment got lost)

File: presto-main/src/main/java/com/facebook/presto/server/ExponentialBackOff.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    public synchronized void success()
+    {
+        lastSuccessTime = ticker.read();
+        failureCount = 0;
+        lastFailureTime = 0;
+    }
+
+    public synchronized boolean failure()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/server/ExponentialBackOff.java]: Document what the return value means

**Comment by martint** [comment]
looks good

**Pull request closed without merging**